Nil pointer when an allocation's task group is no longer found on the job

### Nomad version

Nomad v0.8.4 (dbee1d7d051619e90a809c23cf7e55750900742a)

### Issue

Nil pointer on all servers when they get raft leadership 

I'm not sure what was done to get into this state, but a nil pointer should never happen :) 

```
Created Deployment: "2039a10e-5984-c912-346c-ac0d700603f5"
Desired Changes for "server": (place 1) (inplace 1) (destructive 0) (stop 0) (migrate 0) (ignore 0) (canary 0)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x99a734]

goroutine 84 [running]:
github.com/hashicorp/nomad/nomad/structs.(*Allocation).NextDelay(0xc43a76db00, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/nomad/structs/structs.go:6024 +0x34
github.com/hashicorp/nomad/scheduler.updateRescheduleTracker(0xc43e17f800, 0xc43a76db00, 0xbed28a80200d1d66, 0x7a9387161, 0x2b96de0)
	/opt/gopath/src/github.com/hashicorp/nomad/scheduler/generic_sched.go:588 +0x331
github.com/hashicorp/nomad/scheduler.(*GenericScheduler).computePlacements(0xc434fe3400, 0x2bb8aa0, 0x0, 0x0, 0xc432b94b00, 0x1, 0x1, 0x1, 0x9)
	/opt/gopath/src/github.com/hashicorp/nomad/scheduler/generic_sched.go:497 +0xd1f
github.com/hashicorp/nomad/scheduler.(*GenericScheduler).computeJobAllocs(0xc434fe3400, 0xc43a68ab60, 0xc434234e40)
	/opt/gopath/src/github.com/hashicorp/nomad/scheduler/generic_sched.go:410 +0x1586
github.com/hashicorp/nomad/scheduler.(*GenericScheduler).process(0xc434fe3400, 0xe, 0xc42598f070, 0x7)
	/opt/gopath/src/github.com/hashicorp/nomad/scheduler/generic_sched.go:245 +0x471
github.com/hashicorp/nomad/scheduler.(*GenericScheduler).(github.com/hashicorp/nomad/scheduler.process)-fm(0xc43e035750, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/scheduler/generic_sched.go:144 +0x2a
github.com/hashicorp/nomad/scheduler.retryMax(0x5, 0xc43e0358b0, 0xc43e0358c0, 0xc, 0xffffffffffffffff)
	/opt/gopath/src/github.com/hashicorp/nomad/scheduler/util.go:271 +0x43
github.com/hashicorp/nomad/scheduler.(*GenericScheduler).Process(0xc434fe3400, 0xc434205b80, 0xc420200b90, 0x1d7cde0)
	/opt/gopath/src/github.com/hashicorp/nomad/scheduler/generic_sched.go:144 +0x123
github.com/hashicorp/nomad/nomad.(*nomadFSM).reconcileQueuedAllocations(0xc420387080, 0x1ca3ae7, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/nomad/fsm.go:1383 +0x8ed
github.com/hashicorp/nomad/nomad.(*nomadFSM).applyReconcileSummaries(0xc420387080, 0xc434302075, 0x8, 0x8, 0x1ca3ae7, 0xc42e5aff76, 0x2b96de0)
	/opt/gopath/src/github.com/hashicorp/nomad/nomad/fsm.go:758 +0x7e
github.com/hashicorp/nomad/nomad.(*nomadFSM).Apply(0xc420387080, 0xc42be382d0, 0x2b96de0, 0x3)
	/opt/gopath/src/github.com/hashicorp/nomad/nomad/fsm.go:210 +0x6f1
github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*Raft).runFSM.func1(0xc43c2275a0)
	/opt/gopath/src/github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/fsm.go:57 +0x15a
github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*Raft).runFSM(0xc42018e000)
	/opt/gopath/src/github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/fsm.go:120 +0x2fa
github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*Raft).(github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.runFSM)-fm()
	/opt/gopath/src/github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/api.go:506 +0x2a
github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc42018e000, 0xc42021a710)
	/opt/gopath/src/github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/state.go:146 +0x53
created by github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*raftState).goFunc
	/opt/gopath/src/github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/state.go:144 +0x66
```

--- a/nomad/structs/structs.go
+++ b/nomad/structs/structs.go
@@ -6293,6 +6293,10 @@ func (a *Allocation) NextRescheduleTime() (time.Time, bool) {
 // It is calculated according to the delay function and previous reschedule attempts.
 func (a *Allocation) NextDelay() time.Duration {
 	policy := a.ReschedulePolicy()
+	// Can be nil if the task group was updated to remove its reschedule policy
+	if policy == nil {
+		return 0
+	}
 	delayDur := policy.Delay
 	if a.RescheduleTracker == nil || a.RescheduleTracker.Events == nil || len(a.RescheduleTracker.Events) == 0 {
 		return delayDur

--- a/nomad/structs/structs_test.go
+++ b/nomad/structs/structs_test.go
@@ -2942,6 +2942,12 @@ func TestAllocation_NextDelay(t *testing.T) {
 			expectedRescheduleEligible: false,
 		},
 		{
+			desc:                       "Allocation has no reschedule policy",
+			alloc:                      &Allocation{},
+			expectedRescheduleTime:     time.Time{},
+			expectedRescheduleEligible: false,
+		},
+		{
 			desc: "Allocation lacks task state",
 			reschedulePolicy: &ReschedulePolicy{
 				DelayFunction: "constant",
@@ -3385,7 +3391,9 @@ func TestAllocation_NextDelay(t *testing.T) {
 		t.Run(tc.desc, func(t *testing.T) {
 			require := require.New(t)
 			j := testJob()
-			j.TaskGroups[0].ReschedulePolicy = tc.reschedulePolicy
+			if tc.reschedulePolicy != nil {
+				j.TaskGroups[0].ReschedulePolicy = tc.reschedulePolicy
+			}
 			tc.alloc.Job = j
 			tc.alloc.TaskGroup = j.TaskGroups[0].Name
 			reschedTime, allowed := tc.alloc.NextRescheduleTime()

